pkg/activator/handler/concurrency_reporter_test.go

```diff
@@ -17,104 +17,299 @@ limitations under the License.
 package handler
 
 import (
+	"sync"
 	"testing"
 	"time"
 
 	"github.com/google/go-cmp/cmp/cmpopts"
 
 	"github.com/google/go-cmp/cmp"
 	"github.com/knative/serving/pkg/autoscaler"
+	"github.com/knative/serving/pkg/system"
 )
 
-func TestMultipleDifferentKeys(t *testing.T) {
+const (
+	requestOpTick  = "RequestOpTick"
+	requestOpStart = "RequestOpStart"
+	requestOpEnd   = "RequestOpEnd"
+)
 
-	pod1 := "pod1"
-	pod2 := "pod2"
+type fakeClock struct {
+	Time time.Time
+}
 
-	s := newTestStats()
+func (c fakeClock) Now() time.Time {
+	return c.Time
+}
 
-	s.requestStart(pod1)
-	s.requestStart(pod1)
-	s.requestStart(pod2)
+type reqOp struct {
+	op   string
+	key  string
+	time time.Time
+}
 
-	now := time.Now()
-	expectStats(t, s.report(now, 2), []*autoscaler.StatMessage{{
-		Key: pod1,
-		Stat: autoscaler.Stat{
-			Time:                      &now,
-			PodName:                   autoscaler.ActivatorPodName,
-			AverageConcurrentRequests: 2.0,
-			RequestCount:              2,
+func TestStats(t *testing.T) {
+	tt := []struct {
+		name          string
+		ops           []reqOp
+		expectedStats []*autoscaler.StatMessage
+	}{
+		{
```

Comment:
You can make this quite a bit more compact by using the `{{` and `}, {` notation. Less lines needed and more importantly: less indentation needed.

Sketched that here for it's a bit hard to explain: https://github.com/knative/serving/commit/31ad9732e27d52ec01be9aa2e0a18b124baa57ed